pkg/egresspolicy/egresspolicy.go

```diff
@@ -0,0 +1,566 @@
+//  limitations under the License.
+
+package egresspolicy
+
+import (
+	"fmt"
+	"net"
+
+	k8sConst "github.com/cilium/cilium/pkg/k8s/apis/cilium.io"
+	v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	k8sUtils "github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/maps/egressmap"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "egresspolicy")
+)
+
+// EgressPolicyConfig is the internal representation of Cilium Egress NAT Policy.
+type EgressPolicyConfig struct {
+	// id is the parsed config name and namespace
+	id types.NamespacedName
+	// uid is the unique identifier assigned by Kubernetes
+	uid types.UID
+
+	endpointSelectors []api.EndpointSelector
+	dstCIDRs          []*net.IPNet
+	egressIP          net.IP
+}
+
+// The egresspolicy manager stores the internal data tracking the policy
+// and pod mappings. And also hooks up all the callbacks to update the egress
+// map accordingly.
+type Manager struct {
+	mutex lock.Mutex
+
+	// Stores pod to policy mapping
+	policyPods map[podID][]policyID
+	// Stores policy configs indexed by policyID
+	policyConfigs map[policyID]*EgressPolicyConfig
+	// Stores podId to pod metadata mapping
+	podDataStore map[podID]*podMetadata
+
+	storeGetter StoreGetter
+}
+
+// podMetadata stores relevant metadata associated with a pod that's updated during pod
+// add/update events
+type podMetadata struct {
+	// Pod labels
+	labels map[string]string
+	// Namespace labels
+	nsLabels map[string]string
+	// id the pod's name and namespace
+	id podID
+	// ips are pod's unique IPs
+	ips []string
+}
+
+// podID is pod name and namespace
+type podID = types.NamespacedName
+
+// PolicyID includes policy name and namespace
+type policyID = types.NamespacedName
+
+type StoreGetter interface {
+	GetStore(name string) cache.Store
+}
+
+// policyConfigSelectsPod determines if the given pod is selected by the policy
+// config based on matching labels of config and pod.
+func (config *EgressPolicyConfig) policyConfigSelectsPod(podInfo *podMetadata) bool {
+	labelsToMatch := labels.Set(podInfo.labels)
+	nsLabels := labels.Set(podInfo.nsLabels)
+	for k, v := range nsLabels {
+		labelsToMatch[policy.JoinPath(k8sConst.PodNamespaceMetaLabels, k)] = v
+	}
+	for _, selector := range config.endpointSelectors {
+		if selector.Matches(labelsToMatch) {
+			return true
+		}
+	}
+	return false
+}
+
+func Parse(cenp *v2alpha1.CiliumEgressNATPolicy) (*EgressPolicyConfig, error) {
```

Comment:
Please add godocs to all exported entities

Reply:
added.